### PR TITLE
Save query history across restarts

### PR DESCRIPTION
**Pull request by aeisenberg**
Successfully completed queries will be stored on disk and available
across restarts.

- The query results are contained in global storage.
- Metadata and a summary about a query are stored in workspace storage.
- There is a job that runs every 2 hours to determine if any queries are
  old enough to be deleted.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/query-history.ts [hunk 1]
```diff
@@ -884,3 +951,106 @@ the file in the file explorer and dragging it into the workspace.`
...
+  queryDirectory: string,
+  ctx: ExtensionContext,
+
+  // optional counter to keep track of how many times the scrubber has run
+  counter?: {
+    increment: () => void;
+  }
+): Disposable {
+  const deregister = setInterval(async () => {
+    const lastScrubTime = ctx.globalState.get<number>(LAST_SCRUB_TIME_KEY);
```
Comment (charisk) [inline extensions/ql-vscode/src/query-history.ts:985]: Do we also want to store the fact that we're currently scrubbing, in global state, and check that before kicking off the process? That would help avoid multiple workspaces running the scrubber at the same time. We'd need to be careful though about reseting that though.

Or are we happy that if two scrubbing processes did happen to run at the same time, we have enough defensiveness that it wouldn't cause trouble?

File: extensions/ql-vscode/src/query-results.ts [hunk 2]
```diff
@@ -253,13 +270,15 @@ export class FullQueryInfo {
...
-    private readonly source?: CancellationTokenSource
+    private source?: CancellationTokenSource // used to cancel in progress queries
   ) {
     this.setConfig(config);
   }
 
   cancel() {
     this.source?.cancel();
+    // query is no longer in progress, can delete the cancellation token source
+    delete this.source;
```
Comment (charisk) [inline extensions/ql-vscode/src/query-results.ts:281]: Why do we need to do this?

File: extensions/ql-vscode/src/run-queries.ts [hunk 3]
```diff
@@ -651,19 +664,21 @@ export async function compileAndRunQueryAgainstDatabase(
...
     hasMetadataFile,
     packConfig.dbscheme,
     initialInfo.quickEvalPosition,
     metadata,
     templates
   );
+  await query.createTimestampFile();
 
-  const upgradeDir = await tmp.dir({ dir: upgradesTmpDir.name, unsafeCleanup: true });
+  let upgradeDir: tmp.DirectoryResult | undefined;
```
Comment (charisk) [inline extensions/ql-vscode/src/run-queries.ts:677]: Is this change relevant to query history? 🤔

File: extensions/ql-vscode/src/query-history.ts [hunk 4]
```diff
@@ -488,19 +540,21 @@ export class QueryHistoryManager extends DisposableObject {
     singleItem: FullQueryInfo,
     multiSelect: FullQueryInfo[]
   ): Promise<void> {
-    if (!this.assertSingleQuery(multiSelect)) {
+    const { finalSingleItem, finalMultiSelect } = this.determineSelection(singleItem, multiSelect);
+
+    if (!this.assertSingleQuery(finalMultiSelect)) {
```
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/query-history.ts:545]: Thanks for pointing it out, and for making the fix! 

In the future, it'd personally help me a lot of we had separate PRs for such changes. I can live with it if it's too much trouble, I just find that I miss stuff more in bigger PRs that do multiple things.

File: extensions/ql-vscode/src/query-history.ts [hunk 5]
```diff
@@ -884,3 +951,106 @@ the file in the file explorer and dragging it into the workspace.`
...
+                void logger.log(`  ${dir} is older than ${maxQueryTime / 1000} seconds. Deleting.`);
+                await fs.remove(dir);
+                scrubCount++;
+              } else {
+                void logger.log(`  ${dir} is not older than ${maxQueryTime / 1000} seconds. Keeping.`);
+              }
+            }
+          } catch (err) {
+            errors.push(`  Could not delete '${dir}': ${err}`);
+          }
```
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/query-history.ts:1034]: That seems like the right place to me!

File: extensions/ql-vscode/src/run-queries.ts [hunk 6]
```diff
@@ -98,6 +94,16 @@ export class QueryEvaluationInfo {
     return path.join(this.querySaveDir, `sortedResults-${resultSetName}.bqrs`);
   }
 
+  /**
+   * Creates a file in the query directory that indicates when this query was created.
+   * This is important for keeping track of when queries should be removed.
+   */
+  async createTimestampFile() {
```
Comment (charisk) [inline extensions/ql-vscode/src/run-queries.ts:101]: Shouldn't this be part of the information we store for the query in the all-history.json?

Happy to be told that that would be too big of a change right now.

File: extensions/ql-vscode/src/query-history.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:545]: In general, I agree and prefer that for reviewing. But for larger changes like these, where nothing is working (until everything is), I find it quite hard to break out these smaller bits, especially since these bits are often somewhat dependent on the larger change. Sometimes, I try separating these out into individual commits in the same PR.

File: extensions/ql-vscode/src/query-history.ts [hunk 7]
```diff
@@ -381,6 +400,16 @@ export class QueryHistoryManager extends DisposableObject {
     this.push(
       queryHistoryConfigListener.onDidChangeConfiguration(() => {
         this.treeDataProvider.refresh();
+        // recreate the history scrubber
+        this.queryHistoryScrubber.dispose();
+        this.queryHistoryScrubber = this.push(
+          registerQueryHistoryScubber(
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:406]: Actually, yes. I should. There are two instances of calling this and they should be identical (currently, they are not).

File: extensions/ql-vscode/src/query-history.ts [hunk 8]
```diff
@@ -398,6 +427,27 @@ export class QueryHistoryManager extends DisposableObject {
...
+    );
+  }
+
+  async readQueryHistory(): Promise<void> {
+    const history = await FullQueryInfo.slurp(this.queryMetadataStorageLocation, this.queryHistoryConfigListener);
+    this.treeDataProvider.allHistory = history;
+  }
+
+  async writeQueryHistory(): Promise<void> {
+    const toSave = this.treeDataProvider.allHistory.filter(q => q.isCompleted());
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:449]: No. Not cancelled queries. It does include queries that are unsuccessful. This allows users to view the logs and find out why the query failed.

File: extensions/ql-vscode/src/query-history.ts [hunk 9]
```diff
@@ -445,14 +495,16 @@ export class QueryHistoryManager extends DisposableObject {
...
-
-    (finalMultiSelect || [finalSingleItem]).forEach((item) => {
-      // Removing in progress queries is not supported yet
+    const toDelete = (finalMultiSelect || [finalSingleItem]);
+    await Promise.all(toDelete.map(async (item) => {
+      // Removing in progress queries is not supported. They must be cancelled first.
       if (item.status !== QueryStatus.InProgress) {
         this.treeDataProvider.remove(item);
         item.completedQuery?.dispose();
+        await item.completedQuery?.query.cleanUp();
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:504]: Two reasons, we can't do this:

1. Dispose is an inherited method and cannot be made async
2. `cleanUp` will delete the entire directory of the query. We don't want to do this on a normal dispose, which gets called when the application exits. We only want to do this when the user explicitly wants to remove the query.

I can add a comment here. And I think `cleanUp` is not the right name. I think `deleteQuery` is better.

File: extensions/ql-vscode/src/query-history.ts [hunk 10]
```diff
@@ -884,3 +951,106 @@ the file in the file explorer and dragging it into the workspace.`
...
+ *
+ * Generally, `wakeInterval` should be significantly shorter than `throttleTime`.
+ *
+ * @param wakeInterval How often to check to see if the job should run.
+ * @param throttleTime How often to actually run the job.
+ * @param maxQueryTime The maximum age of a query before is ready for deletion.
+ * @param queryDirectory The directory containing all queries.
+ * @param ctx The extension context.
+ */
+export function registerQueryHistoryScubber(
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:972]: Done.

File: extensions/ql-vscode/src/query-history.ts (hunk 1 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:985]: I think a safe and easy thing to do is to move the `ctx.globalState.update` to happen as soon as possible after the `get`. This minimizes the window when two workspaces would be able to run the scrubber at the same time. Fundamentally though, I don't think this would be a problem. There might be some extra error messages in one of the workspaces if it gets a list of query directories, but by the time it checks for them they are gone.

File: extensions/ql-vscode/src/query-results.ts (hunk 2 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-results.ts:281]: The `source` object is complex and contains workspace state. I'm doing it to make sure workspace resources aren't leaked. I realized that this object is a `Disposable` as well, so I'll add a call to `dispose`.

Slightly unrelated, this object can't be serialized to JSON because it's recursive, so it also needs to be deleted before serialization.

File: extensions/ql-vscode/src/run-queries.ts (hunk 6 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/run-queries.ts:101]: It can't be stored there. We need to store this value in the query directory so that all workspaces know how old a given query folder is.

Remember, that the goal is to be able to clean up queries for workspaces that you never intend to open again.

File: extensions/ql-vscode/src/contextual/templateProvider.ts [hunk 11]
```diff
@@ -42,6 +42,7 @@ export class TemplateQueryDefinitionProvider implements DefinitionProvider {
     private cli: CodeQLCliServer,
     private qs: QueryServerClient,
     private dbm: DatabaseManager,
+    private queryStorageLocation: string,
```
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/contextual/templateProvider.ts]: Yep that works

File: extensions/ql-vscode/src/query-history-scrubber.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  const deregister = setInterval(scrubber, wakeInterval, throttleTime, maxQueryTime, queryDirectory, ctx, counter);
+
+  return {
+    dispose: () => {
+      clearInterval(deregister);
+    }
+  };
+}
+
+async function scrubber(
```
Comment (charisk) [inline extensions/ql-vscode/src/query-history-scrubber.ts:47]: How about calling this `scrub` instead (or `scrubHistory`)? I generally try to use verbs for functions.

File: extensions/ql-vscode/src/query-history.ts (hunk 1 above)
Comment (charisk) [reply extensions/ql-vscode/src/query-history.ts]: Ok cool - we can see how it goes in practise too and adjust if necessary.

File: extensions/ql-vscode/src/query-results.ts (hunk 2 above)
Comment (charisk) [reply extensions/ql-vscode/src/query-results.ts]: That sounds fun 😱

File: extensions/ql-vscode/src/run-queries.ts (hunk 6 above)
Comment (charisk) [reply extensions/ql-vscode/src/run-queries.ts:101]: > It can't be stored there. We need to store this value in the query directory so that all workspaces know how old a given query folder is.
Is all-history.json not meant to be stored in global storage?

File: extensions/ql-vscode/src/query-history.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (charisk) [reply extensions/ql-vscode/src/query-history.ts:553]: I think it's still ok to separate things out in separate PRs even when you work on a larger bit of work. 

If I'm working on something fairly big I start off the work and if I find something that could be split into its own PR then I do that. It's a bit more hassle juggling things but I find that 
a) it's easier for the reviewer 
b) things get done more quickly overall and 
c) there's less of a chance that things get missed

I'm probably "preaching to the converted" and apologies for that, just wanted to make it clear what my personal preferences (and weaknesses!) are as a reviewer.

File: extensions/ql-vscode/src/query-history-scrubber.ts (hunk 12 above)
Comment (charisk) [inline extensions/ql-vscode/src/query-history-scrubber.ts:47]: How about calling this scrub instead (or scrubHistory)? I generally try to use verbs for functions.

File: extensions/ql-vscode/src/run-queries.ts (hunk 6 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/run-queries.ts:101]: That's correct. `all-history.json` is in workspace storage. This way each workspace only sees the queries relevant to itself. Other workspaces' queries will not show.

I can change the name and be more clear.

**Review by charisk — APPROVED** [review]
